lib/io/stdin.fz

```diff
@@ -35,14 +35,46 @@ private stdin(
   ) : stdins, effect r is
 
 
-  # for a stdin instance installed in the current environment,
-  # install next instance of byteInputHandler,
-  # 'stdin.env' will provide the new instance.
-  #
-  private next => stdin ip.next mode unit
-
-  nextByte =>
-    next.ip.get
+  # read a single byte from stdin
+  byte =>
+    (stdin ip.next mode unit).ip.get
+
+
+  # an infinite list of either a byte, an end_of_file or an error
+  # Caution!: this list is not immutable since it is constructed lazily by reading from stdin!
+  # you may iterate only once.
+  # Additionally there may be read more from stdin than what is consumed
+  # via this list.
+  # NYI: How can we lessen these imperfections?
+  bytes list (choice u8 end_of_file error) is
+    list byte ()->bytes
+
+
+  # try reading a codepoint from stdin
+  codepoint =>
+    read_codepoint(read_bytes list u8) choice error end_of_file codepoint is
+      match byte
+        byte u8 =>
+          bytes := read_bytes ++ [byte]
+          match (strings.fromBytes bytes).codepointsAndErrors.first
+            c codepoint => c
+            e error => if bytes.count >= 4 then e else read_codepoint bytes
```

Comment:
instead of trying 4 times and checking `count >= 4`, it would IMHO be better to add a feature `codepoints.num_utf8_bytes(first_byte u8)` that returns `1` for 0..0x7f, `2`for `0b110xxxxx`, `3` for `0b1110xxxx`, and `4` for `0b11110xxx`. 
(In a separate patch,) You could even add `u8.leadingZeroes` and implement `num_utf8_bytes(first_byte u8) => (~first_byte).leadingZeroes+1`.